src/plugins/collapsibleColumns/collapsibleColumns.js

```diff
@@ -442,118 +267,118 @@ class CollapsibleColumns extends BasePlugin {
 
-      rangeEach(firstChildColspan, currentHeaderColspan - 1, (i) => {
-        const colToHide = currentCoords.col + i;
+    if (isActionPossible) {
+      arrayEach(filteredCoords, ({ row, col: column }) => {
+        const {
+          colspanCompensation,
+          affectedColumns,
+          rollbackModification,
+        } = this.headerStateManager.triggerNodeModification(action, row, column);
 
-        switch (action) {
-          case 'collapse':
-            if (!this.hiddenColumnsPlugin.isHidden(colToHide)) {
-              this.hiddenColumnsPlugin.hideColumn(colToHide);
-            }
+        if (colspanCompensation > 0) {
+          affectedColumnsIndexes.push(...affectedColumns);
+          nodeModRollbacks.push(rollbackModification);
+        }
+      });
+    }
 
-            this.markSectionAs('collapsed', currentCoords.row, currentCoords.col, true);
+    const currentCollapsedColumns = [...this.hiddenColumnsPlugin.getHiddenColumns()];
+    let destinationCollapsedColumns = [];
 
-            break;
-          case 'expand':
-            if (this.hiddenColumnsPlugin.isHidden(colToHide)) {
-              this.hiddenColumnsPlugin.showColumn(colToHide);
-            }
+    if (action === 'collapse') {
+      destinationCollapsedColumns = arrayUnique([...currentCollapsedColumns, ...affectedColumnsIndexes]);
 
-            this.markSectionAs('expanded', currentCoords.row, currentCoords.col, true);
+    } else if (action === 'expand') {
+      destinationCollapsedColumns = arrayFilter(currentCollapsedColumns, index => !affectedColumnsIndexes.includes(index));
+    }
 
-            break;
-          default:
-            break;
-        }
+    const actionTranslator = actionDictionary.get(action);
+    const isActionAllowed = this.hot.runHooks(
+      actionTranslator.beforeHook,
+      currentCollapsedColumns,
+      destinationCollapsedColumns,
+      isActionPossible,
+    );
+
+    if (isActionAllowed === false) {
+      // Rollback all header nodes modification (collapse or expand).
+      arrayEach(nodeModRollbacks, (nodeModRollback) => {
+        nodeModRollback();
       });
-    });
-
-    // Added by @wszymanski, this code around was absent, when the original commit was performed.
-    const destinationCollapsedColumns = this.hiddenColumnsPlugin.getHiddenColumns();
 
-    if (action === 'collapse') {
-      const allowColumnCollapse = this.hot.runHooks('beforeColumnCollapse', currentCollapsedColumns, destinationCollapsedColumns, collapsePossible);
+      return;
+    }
 
-      if (allowColumnCollapse === false) {
-        // Added by @wszymanski, this code around was absent, when the original commit was performed.
-        this.hiddenColumnsPlugin.showColumns(this.hiddenColumnsPlugin.getHiddenColumns());
-        this.hiddenColumnsPlugin.hideColumns(this.collapsedColumns);
-        this.collapsedSections = cloneCollapsedSections;
+    this.hiddenColumnsPlugin[actionTranslator.hiddenAction](affectedColumnsIndexes);
```

Comment:
It will be refactored when the `CollapsibleColumns` plugin will be refactored to use just the `HidingMap` 🤔